pkg/emb6/contrib/sock/udp/emb6_sock_udp.c

```diff
@@ -0,0 +1,323 @@
+        timeout_timer.arg = &sock->mbox;
+        xtimer_set(&timeout_timer, timeout);
+    }
+    atomic_fetch_add(&sock->receivers, 1);
+    if (_mbox_get(&sock->mbox, &msg, blocking) == 0) {
+        /* do not need to remove xtimer, since we only get here in non-blocking
+         * mode (timeout > 0) */
+        return -EAGAIN;
+    }
+    switch (msg.type) {
+        case _MSG_TYPE_CLOSE:
+            res = -EADDRNOTAVAIL;
+            break;
+        case _MSG_TYPE_TIMEOUT:
+            res = -ETIMEDOUT;
+            break;
+        case _MSG_TYPE_RCV:
+            mutex_lock(&sock->mutex);
+            if (max_len < sock->recv_info.datalen) {
+                res = -ENOBUFS;
+                mutex_unlock(&sock->mutex);
+                break;
+            }
+            memcpy(data, sock->recv_info.data, sock->recv_info.datalen);
+            if (remote != NULL) {
+                remote->family = AF_INET6;
+                remote->netif = SOCK_ADDR_ANY_NETIF;
+                memcpy(&remote->addr, &sock->recv_info.src, sizeof(ipv6_addr_t));
+                remote->port = sock->recv_info.src_port;
+            }
+            res = (int)sock->recv_info.datalen;
+            mutex_unlock(&sock->mutex);
+            break;
+    }
+    atomic_fetch_sub(&sock->receivers, 1);
+    return res;
+}
+
+int sock_udp_send(sock_udp_t *sock, const void *data, size_t len,
+                  const sock_udp_ep_t *remote)
+{
+    struct udp_socket tmp;
+    _send_cmd_t send_cmd = { .block = MUTEX_INIT,
+                             .remote = remote,
+                             .data = data,
+                             .len = len };
+
+    assert((sock != NULL) || (remote != NULL));
+    assert((len == 0) || (data != NULL));   /* (len != 0) => (data != NULL) */
+    /* we want the send in the uip thread (which udp_socket_send does not offer)
+     * so we need to do it manually */
+    if (!send_registered) {
+        if (evproc_regCallback(EVENT_TYPE_CONN_SEND, _output_callback) != E_SUCCESS) {
+            return -ENOMEM;
+        }
+        else {
+            send_registered = true;
+        }
+    }
+    if ((len > (UIP_BUFSIZE - (UIP_LLH_LEN + UIP_IPUDPH_LEN))) ||
+        (len > UINT16_MAX)) {
+        return -ENOMEM;
+    }
+    if ((remote != NULL)) {
+        if (remote->family != AF_INET6) {
+            return -EAFNOSUPPORT;
+        }
+        if (remote->port == 0) {
+            return -EINVAL;
+        }
+        send_cmd.remote = remote;
+    }
+    if ((remote == NULL) && (sock->sock.udp_conn->rport == 0)) {
+        return -ENOTCONN;
+    }
+    /* cppcheck-suppress nullPointerRedundantCheck
+     * remote == NULL implies that sock != NULL (see assert at start of functon)
+     * that's why it is okay in the if-statement above to check sock->...
+     * without checking (sock != NULL) first => this check afterwards isn't
+     * redundant */
+    if (sock == NULL) {
+        int res;
```

Comment:
Because it's the result of the send command, not the result of `_reg()` (and I don't want to dig into emb6's code now to make sure I don't have to zero it again bevore puttilng `send_cmd` in its event queue ;-))